### PR TITLE
Add hook for opendir, issue #1899

### PR DESCRIPTION
**Pull request by meowjesty**
- Issue: #1899

Add hook for `opendir` by leveraging our `open` and `fdopendir` hooks.

Also adds hooks for `readdir64_r`, `__lxstat`, `__lxstat64`, `__xstat64`.

Refactors `*stat` hooks to use a common function.

Adds an integration test for `opendir`.

### CODE REVIEW

File: mirrord/layer/src/file/hooks.rs [hunk 1]
```diff
@@ -133,6 +133,17 @@ pub(super) unsafe extern "C" fn open_nocancel_detour(
     }
 }
 
+/// Hook for [`libc::opendir`].
+///
+/// Opens the directory with `read` permission using the [`open_logic`] flow, then calls
+/// [`fdopendir`] to convert the [`RawFd`] into a `*DIR` stream (which we treat as `usize`).
+#[hook_guard_fn]
+pub(super) unsafe extern "C" fn opendir_detour(raw_filename: *const c_char) -> usize {
+    open_logic(raw_filename, O_RDONLY, 0)
```
Comment (aviramha) [inline mirrord/layer/src/file/hooks.rs:142]: need to handle case open succeeds but fdopendir fails (close fd)

Comment (aviramha) [reply mirrord/layer/src/file/hooks.rs:142]: I'd add O_DIRECTORY just to be on the safe side (though we ignore it, but maybe we wont in the future)

**Comment by aviramha** [comment]
The integration tests pass for me locally when I comment out the "macOS" part - maybe it's x64 thingy though (needing the xstat?)

**Comment by aviramha** [comment]
> The integration tests pass for me locally when I comment out the "macOS" part - maybe it's x64 thingy though (needing the xstat?)

ah I think I know what's the problem - you need to hook `opendir$INODE64` on macOS x64.

**Comment by aviramha** [comment]
After few tests, I found out that for listing files in dir we lack `readdir64_r`
and after that it still won't work - The best way to trigger the same issue is to check if `Directory.Exists` works - and it doesn't :( - I think it might be using ~stat~ func for it.

**Comment by aviramha** [comment]
[patch.patch](https://github.com/metalbear-co/mirrord/files/12523704/patch.patch)
adding patch file for what I did so far

File: mirrord/layer/src/file/hooks.rs [hunk 2]
```diff
@@ -133,6 +137,17 @@ pub(super) unsafe extern "C" fn open_nocancel_detour(
     }
 }
 
+/// Hook for [`libc::opendir`].
+///
+/// Opens the directory with `read` permission using the [`open_logic`] flow, then calls
+/// [`fdopendir`] to convert the [`RawFd`] into a `*DIR` stream (which we treat as `usize`).
+#[hook_guard_fn]
+pub(super) unsafe extern "C" fn opendir_detour(raw_filename: *const c_char) -> usize {
+    open_logic(raw_filename, O_RDONLY, O_DIRECTORY)
```
Comment (aviramha) [inline mirrord/layer/src/file/hooks.rs:146]: if open succeeds but fdopendir fails, we need to call `close_layer_fd`

**Comment by meowjesty** [comment]
```rust
fn note(&mut self) {
    println!("Remember to update the changelog, as we did more than just add opendir.");
}
```

File: mirrord/layer/src/file/hooks.rs [hunk 3]
```diff
@@ -804,6 +899,7 @@ pub(crate) unsafe fn enable_file_hooks(hook_manager: &mut HookManager) {
     );
 
     replace!(hook_manager, "openat", openat_detour, FnOpenat, FN_OPENAT);
+    replace!(hook_manager, "openat64", openat_detour, FnOpenat, FN_OPENAT);
```
Comment (aviramha) [inline mirrord/layer/src/file/hooks.rs:902]: It'd be more preferable to have different detours, for case where `openat` resolves but `openat64` doesn't (as it assigns it to the `FN_OPENAT` (we had cases like this, that's why inode hooks are in cfg, which tbh we should just have similar logic)

File: mirrord/layer/src/file/hooks.rs (hunk 2 above)
Comment (meowjesty) [reply mirrord/layer/src/file/hooks.rs:146]: Nice catch.

File: mirrord/layer/src/file/hooks.rs (hunk 3 above)
Comment (meowjesty) [reply mirrord/layer/src/file/hooks.rs]: Changed.

**Review by aviramha — APPROVED** [review]

**Pull request merged**